File: app/src/main/java/com/pr0gramm/app/ui/ZoomViewActivity.java
package com.pr0gramm.app.ui;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.ColorRes;
import android.view.View;
import android.widget.ImageView;

import com.akodiakson.sdk.simple.Sdk;
import com.davemorrissey.labs.subscaleview.ImageSource;
import com.davemorrissey.labs.subscaleview.SubsamplingScaleImageView;
import com.f2prateek.dart.InjectExtra;
import com.google.common.base.Strings;
import com.pr0gramm.app.ActivityComponent;
import com.pr0gramm.app.BuildConfig;
import com.pr0gramm.app.R;
import com.pr0gramm.app.Settings;
import com.pr0gramm.app.feed.FeedItem;
import com.pr0gramm.app.services.ThemeHelper;
import com.pr0gramm.app.services.UriHelper;
import com.pr0gramm.app.services.proxy.ProxyService;
import com.pr0gramm.app.ui.base.BaseAppCompatActivity;
import com.pr0gramm.app.util.decoders.Decoders;
import com.pr0gramm.app.util.decoders.PicassoDecoder;
import com.squareup.picasso.Downloader;
import com.squareup.picasso.Picasso;

import javax.inject.Inject;

import butterknife.BindView;
import rx.Emitter;
import rx.Observable;

import static com.pr0gramm.app.services.ThemeHelper.theme;
import static com.pr0gramm.app.util.AndroidUtility.getTintentDrawable;

public class ZoomViewActivity extends BaseAppCompatActivity {
    private final String tag = "ZoomViewActivity" + System.currentTimeMillis();

    @InjectExtra("ZoomViewActivity__item")
    FeedItem item;

    @InjectExtra("ZoomViewActivity__hq")
    boolean loadHqIfAvailable;

    @BindView(R.id.image)
    SubsamplingScaleImageView imageView;

    @BindView(R.id.busy_indicator)
    View busyIndicator;

    @BindView(R.id.hq)
    ImageView hq;

    @Inject
    Picasso picasso;

    @Inject
    Downloader downloader;

    @Inject
    Settings settings;

    @Inject
    ProxyService proxyService;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setTheme(theme().fullscreen);
        super.onCreate(savedInstanceState);

        // normal content view
        setContentView(R.layout.activity_zoom_view);

        imageView.setMaxTileSize(4096);
        imageView.setDebug(BuildConfig.DEBUG);
        imageView.setBitmapDecoderFactory(() -> new PicassoDecoder(tag, picasso));
        imageView.setRegionDecoderFactory(() -> Decoders.newFancyRegionDecoder(downloader));

        rxImageLoaded(imageView)
                .compose(bindToLifecycle())
                .subscribe(event -> {
                    hideBusyIndicator();

                    imageView.setMaxScale(Math.max(
                            2 * (float) imageView.getWidth() / imageView.getSWidth(),
                            2 * (float) imageView.getHeight() / imageView.getSWidth()
                    ));
                });

        hq.setImageDrawable(getColoredHqIcon(R.color.grey_700));

        if (loadHqIfAvailable && isHqImageAvailable()) {
            loadHqImage();
        } else {
            loadImage();
        }
    }

    @Override
    protected void injectComponent(ActivityComponent appComponent) {
        appComponent.inject(this);
    }

    @SuppressLint("InlinedApi")
    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);

        if (hasFocus && settings.fullscreenZoomView()) {
            View decorView = getWindow().getDecorView();
            int flags = View.SYSTEM_UI_FLAG_HIDE_NAVIGATION;

            if (Sdk.isAtLeastJellyBean()) {
                flags |= View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                        | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                        | View.SYSTEM_UI_FLAG_FULLSCREEN;
            }

            if (Sdk.isAtLeastKitKat())
                flags |= View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY;

            decorView.setSystemUiVisibility(flags);
        }
    }

    @Override
    protected void onDestroy() {
        picasso.cancelTag(tag);
        super.onDestroy();
    }

    public static Intent newIntent(Context context, FeedItem item, boolean hq) {
        Intent intent = new Intent(context, ZoomViewActivity.class);
        intent.putExtra("ZoomViewActivity__item", item);
        intent.putExtra("ZoomViewActivity__hq", hq);
        return intent;
    }

    private void loadImage() {
        Uri url = UriHelper.of(this).media(item);
        loadImageWithUrl(url);

        if (isHqImageAvailable()) {
            hq.setOnClickListener(v -> loadHqImage());
            hq.animate().alpha(1).start();
        } else {
            hq.setVisibility(View.GONE);
        }
    }

    private boolean isHqImageAvailable() {
        return !Strings.isNullOrEmpty(item.fullsize());
    }

    private void loadHqImage() {
        hq.setOnClickListener(null);
        hq.setImageDrawable(getColoredHqIcon(ThemeHelper.accentColor()));
        hq.animate().alpha(1).start();

        Uri url = UriHelper.of(this).media(item, true);
        loadImageWithUrl(url);
    }

    private Drawable getColoredHqIcon(@ColorRes int colorId) {
        return getTintentDrawable(this, R.drawable.ic_action_high_quality, colorId);
    }

    private void loadImageWithUrl(Uri url) {
        showBusyIndicator();
        picasso.cancelTag(tag);
        imageView.setImage(ImageSource.uri(url));
    }

    private void showBusyIndicator() {
        if (busyIndicator != null) {
            busyIndicator.setVisibility(View.VISIBLE);
        }
    }

    private void hideBusyIndicator() {
        if (busyIndicator == null) {
            return;
        }

        busyIndicator.setVisibility(View.GONE);
    }

    private static Observable<Void> rxImageLoaded(SubsamplingScaleImageView view) {
        return Observable.create((emitter) -> {
            view.setOnImageEventListener(new SubsamplingScaleImageView.DefaultOnImageEventListener() {
                @Override
                public void onImageLoaded() {
                    emitter.onNext(null);
                    emitter.onCompleted();
                }
            });
        }, Emitter.BackpressureMode.NONE);
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/ZoomViewActivity.kt
package com.pr0gramm.app.ui

import android.content.Context
import android.content.Intent
import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.view.View.SYSTEM_UI_FLAG_IMMERSIVE
import android.widget.ImageView
import androidx.annotation.ColorRes
import androidx.core.view.isVisible
import com.davemorrissey.labs.subscaleview.ImageSource
import com.davemorrissey.labs.subscaleview.SubsamplingScaleImageView
import com.pr0gramm.app.BuildConfig
import com.pr0gramm.app.R
import com.pr0gramm.app.Settings
import com.pr0gramm.app.feed.FeedItem
import com.pr0gramm.app.io.Cache
import com.pr0gramm.app.parcel.getExtraParcelableOrThrow
import com.pr0gramm.app.services.ThemeHelper
import com.pr0gramm.app.services.Track
import com.pr0gramm.app.services.UriHelper
import com.pr0gramm.app.ui.base.BaseAppCompatActivity
import com.pr0gramm.app.util.AndroidUtility.getTintedDrawable
import com.pr0gramm.app.util.decoders.Decoders
import com.pr0gramm.app.util.decoders.PicassoDecoder
import com.pr0gramm.app.util.di.instance
import com.squareup.picasso.Picasso
import kotterknife.bindView
import kotlin.math.max
import kotlin.math.min

class ZoomViewActivity : BaseAppCompatActivity("ZoomViewActivity") {
    private val tag = "ZoomViewActivity" + System.currentTimeMillis()

    internal val item: FeedItem by lazy {
        intent.getExtraParcelableOrThrow("ZoomViewActivity__item")
    }

    private val hq: ImageView by bindView(R.id.hq)

    private val busyIndicator: View by bindView(R.id.busy_indicator)
    private val imageView: SubsamplingScaleImageView by bindView(R.id.image)

    private val picasso: Picasso by instance()
    private val cache: Cache by instance()

    override fun onCreate(savedInstanceState: Bundle?) {
        setTheme(ThemeHelper.theme.fullscreen)
        super.onCreate(savedInstanceState)

        // normal content view
        setContentView(R.layout.activity_zoom_view)

        Track.openZoomView(item.id)

        imageView.setMaxTileSize(2048)
        imageView.setDebug(BuildConfig.DEBUG)
        imageView.setBitmapDecoderFactory(PicassoDecoder.factory(tag, picasso))
        imageView.setRegionDecoderFactory(Decoders.regionDecoderFactory(cache))

        imageView.setOnImageEventListener(object : SubsamplingScaleImageView.DefaultOnImageEventListener() {
            override fun onImageLoaded() {
                hideBusyIndicator()

                imageView.minScale = min(
                        0.5f * imageView.width.toFloat() / imageView.sWidth,
                        0.5f * imageView.height.toFloat() / imageView.sWidth)

                imageView.setMinimumScaleType(SubsamplingScaleImageView.SCALE_TYPE_CUSTOM)

                imageView.maxScale = max(
                        2 * imageView.width.toFloat() / imageView.sWidth,
                        2 * imageView.height.toFloat() / imageView.sWidth)
            }
        })


        if (Settings.loadHqInZoomView && isHqImageAvailable) {
            // dont show the hq button if we load the hq image directly
            hq.isVisible = false

            loadHqImage()
        } else {
            hq.setImageDrawable(getColoredHqIcon(R.color.grey_700))

            loadNormalImage()
        }

        imageView.setOnClickListener {
            toggleSystemUI()
        }

        // set the correct flags on startup
        showSystemUI()
    }

    private fun hideSystemUI() {
        window.decorView.systemUiVisibility = (SYSTEM_UI_FLAG_IMMERSIVE
                // Set the content to appear under the system bars so that the
                // content doesn't resize when the system bars hide and show.
                or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                // Hide the nav bar and status bar
                or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_FULLSCREEN)
    }

    private fun showSystemUI() {
        window.decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN)
    }

    private fun toggleSystemUI() {
        val immersive = window.decorView.systemUiVisibility and SYSTEM_UI_FLAG_IMMERSIVE == 0
        if (immersive) {
            hideSystemUI()
        } else {
            showSystemUI()
        }
    }

    override fun onDestroy() {
        picasso.cancelTag(tag)
        super.onDestroy()
    }

    private fun loadNormalImage() {
        val url = UriHelper.of(this).media(item)
        loadImageWithUrl(url)

        if (isHqImageAvailable) {
            hq.setOnClickListener { loadHqImage() }
            hq.animate().alpha(1f).start()
        } else {
            hq.isVisible = false
        }
    }

    private val isHqImageAvailable: Boolean
        get() = item.fullsize.isNotBlank()

    private fun loadHqImage() {
        if (hq.isVisible) {
            hq.setOnClickListener(null)
            hq.setImageDrawable(getColoredHqIcon(ThemeHelper.accentColor))
            hq.animate().alpha(1f).start()
        }

        val url = UriHelper.of(this).media(item, true)
        loadImageWithUrl(url)
    }

    private fun getColoredHqIcon(@ColorRes colorId: Int): Drawable {
        return getTintedDrawable(this, R.drawable.ic_action_high_quality, colorId)
    }

    private fun loadImageWithUrl(url: Uri) {
        showBusyIndicator()
        picasso.cancelTag(tag)
        imageView.setImage(ImageSource.uri(url))
    }

    private fun showBusyIndicator() {
        busyIndicator.isVisible = true
    }

    private fun hideBusyIndicator() {
        busyIndicator.isVisible = false
    }

    companion object {
        fun newIntent(context: Context, item: FeedItem): Intent {
            val intent = Intent(context, ZoomViewActivity::class.java)
            intent.putExtra("ZoomViewActivity__item", item)
            return intent
        }
    }
}
